libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt

```diff
@@ -0,0 +1,296 @@
+ *     }
+ *
+ *     override fun hashCode(): Int {
+ *         var result = required.hashCode()
+ *         result = 31 * result + foo.hashCode()
+ *         result = 31 * result + (bar ?: 0)
+ *         return result
+ *     }
+ *
+ *     override fun toString(): String {
+ *         return "Example(required='$required', foo='$foo', bar=$bar)"
+ *     }
+ *
+ *     /**
+ *      * Description for when to use this builder
+ *      *
+ *      * @param required does not have a default value
+ *      */
+ *     class Builder(
+ *         private val required: String
+ *     ) {
+ *         private var foo: String = "default foo value"
+ *         private var bar: Int? = null
+ *
+ *         /**
+ *          * @param foo description of what to expect when writing foo
+ *          * @return Builder
+ *          */
+ *         fun foo(foo: String) =
+ *             apply { this.foo = foo }
+ *
+ *         /**
+ *          * @param bar description of what to expect when writing bar
+ *          * @return Builder
+ *          */
+ *         fun bar(bar: Int?) =
+ *             apply { this.bar = bar }
+ *
+ *         /**
+ *          * Build new instance of [Example]
+ *          * @return Example
+ *          */
+ *         fun build(): Example {
+ *             return Example(
+ *                 required,
+ *                 foo,
+ *                 bar
+ *             )
+ *         }
+ *     }
+ * }
+ * ```
+ */
+abstract class BuilderTest<Implementation : Any, Builder> {
+
+    private val implClass: KClass<*> by lazy { getImplementationClass() }
+
+    /**
+     * Your builder's resulting type.
+     */
+    abstract fun getImplementationClass(): KClass<Implementation>
+
+    /**
+     * Make sure that all of the required and optional values are provided and that they distinct from default ones.
+     */
+    abstract fun getFilledUpBuilder(): Builder
+
+    private val toBuilderMethod = implClass.members.find { it.name == "toBuilder" } as? KFunction
+        ?: throw RuntimeException("missing toBuilder method")
+
+    private val builderClass = implClass.nestedClasses.find { it.simpleName == "Builder" }
+        ?: throw RuntimeException("missing Builder nested class")
+    private val requiredFieldNames =
+        builderClass.members.filter { it is KProperty && it !is KMutableProperty }.map { it.name }
+    private val optionalFieldNames =
+        builderClass.members.filter { it is KProperty && it is KMutableProperty }.map { it.name }
+    private val buildMethod = builderClass.members.find { it.name == "build" } as? KFunction
+        ?: throw RuntimeException("missing Builder.build method")
+
+    /**
+     * This method needs to be overridden and HAS TO explicitly add the @Test annotation in the child.
+     * This is needed to trick JUnit4 to process the test class if all the actual test cases are in the abstract parent.
+     * TODO add a lint rule to detect the missing annotation in the child, or create a new test runner
```

Comment:
What if this class was a TestRule? we could potentially eliminate the need for this

Reply:
I believe test rules can only perform actions before and after the test, they cannot run actual test cases. Also, they are applied before the test setup, so it would be too early to execute any assertions if there's a need for preparation.